File: Graphs/DetectCyclesIn2DGrid.py
# LeetCode 1559 - Detect Cycles in 2D Grid
# Time Complexity: O(m * n) | Space Complexity: O(m * n)
from typing import List


class Solution:
    def containsCycle(self, grid: List[List[str]]) -> bool:
        rows = len(grid)
        cols = len(grid[0])
        visited = [[False] * cols for _ in range(rows)]
        directions = (-1, 0, 1, 0, -1)

        def dfs(row: int, col: int, parent_row: int, parent_col: int) -> bool:
            visited[row][col] = True

            for index in range(4):
                next_row = row + directions[index]
                next_col = col + directions[index + 1]

                if not (0 <= next_row < rows and 0 <= next_col < cols):
                    continue

                if grid[next_row][next_col] != grid[row][col]:
                    continue

                if next_row == parent_row and next_col == parent_col:
                    continue

                if visited[next_row][next_col] or dfs(next_row, next_col, row, col):
                    return True

            return False

        for row in range(rows):
            for col in range(cols):
                if not visited[row][col] and dfs(row, col, -1, -1):
                    return True

        return False

File: Graphs/DetectCyclesIn2DGrid.java
// LeetCode 1559 - Detect Cycles in 2D Grid
// Time Complexity: O(m * n) | Space Complexity: O(m * n)
public class DetectCyclesIn2DGrid {
    private char[][] grid;
    private boolean[][] visited;
    private final int[] directions = {-1, 0, 1, 0, -1};
    private int rows;
    private int cols;

    public boolean containsCycle(char[][] grid) {
        this.grid = grid;
        rows = grid.length;
        cols = grid[0].length;
        visited = new boolean[rows][cols];

        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if (!visited[row][col] && dfs(row, col, -1, -1)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean dfs(int row, int col, int parentRow, int parentCol) {
        visited[row][col] = true;

        for (int index = 0; index < 4; index++) {
            int nextRow = row + directions[index];
            int nextCol = col + directions[index + 1];

            if (nextRow < 0 || nextRow >= rows || nextCol < 0 || nextCol >= cols) {
                continue;
            }

            if (grid[nextRow][nextCol] != grid[row][col]
                    || (nextRow == parentRow && nextCol == parentCol)) {
                continue;
            }

            if (visited[nextRow][nextCol] || dfs(nextRow, nextCol, row, col)) {
                return true;
            }
        }

        return false;
    }
}
